keps/sig-scheduling/20200116-coscheduling-plugin-based-on-scheduler-framework.md

```diff
@@ -0,0 +1,216 @@
+   
+4. When podA and podB are both pgPods. Compare the the `LastFailureTimestamp` of the podGroup. 
+   1. If `LastFailureTimestampA` is earlier than `LastFailureTimestampB`, return true. `LastFailureTimestampA` is later than `LastFailureTimestampB`, return false.
+   
+   2. If `LastFailureTimestampA` is equal to `LastFailureTimestampB`, we will compare the `UID` of `PodGroup`. (The purpose is to distinguish different `PodGroup` with the same `LastFailureTimestamp` and to keep the pods of the same `PodGroup` together)
+
+**Note1**: There are different `LastFailureTimestamp` (even if they are the same, the UID will be different). So when the pods enter the queue, the pods that belongs to the same PodGroup will be together.
+
+#### Pre-Filter
+1. When a pod which belongs to a `PodGroup` comes, calculate the total number of Pods  excludes terminating ones belongs to the same PodGroup. If the result is less than minAvailable, the scheduling process is terminated directly.
+   
+2. Compare whether the `PodGroup` of the current pod is `LastPodGroup`, and if so, keep it unchanged. If not, it indicates that the new PodGroup scheduling cycle has been entered. Update LastPodGroup. And if there is a pod belongs to the last podgroup is waiting and if no chance to meet minAvailable, reject it directly here. The purpose is 1. To avoid invalid waiting, 2. To make the pod belongs to the same `PodGroup` fail together, rather than waiting partially.
+
+```go
+func (cs *Coscheduling) PreFilter(ctx context.Context, state *framework.CycleState, p *v1.Pod) *framework.Status {
+    …
+ 
+     // Count the numbers of all pods belongs to the same PodGroup excludes terminating ones.
+    total := cs.calculateTotalPods(podGroupName)
+    if total < minAvailable {
+        return framework.NewStatus(framework.Unschedulable, "less than minAvailable")
+    }
+    ...
+    if cs.Last != "" && cs.Last != podGroupName {
+        allocated := cs.calculateAllocatedPods(cs.LastPodGroup)
+        pgInfo, ok := cs.PodGroupInfos[cs.LastPodGroup]
+        if ok {
+        	pgInfo.LastFailureTimestamp = time.Now()
+        	if allocated < pgInfo.minAvailable {
+            	cs.FrameworkHandle.IterateOverWaitingPods(func(p framework.WaitingPod) {
+                	if p.GetPod().Labels[PodGroupName] == cs.LastPodGroup {
+                    	p.Reject(cs.Name())
+                	}
+            	})
+        	}
+        }
+    }
+    cs.LastPodGroup = podGroupName
+    return nil
+}
+```
+
+#### Permit
+1. Get the number of Running pods that belong to the same PodGroup
+2. Get the number of WaitingPods (used to record pods in waiting status) that belong to the same PodGroup
+3. If Running + WaitingPods + 1 >= min-available(1 means the pod itself), apprve the waiting pods that  belongs to the same PodGroup. Otherwise, put the pod into WaitingPods and set the timeout(eg: 10s)
```

Comment:
```suggestion
3. If Running + WaitingPods + 1 >= min-available(1 means the pod itself), approve the waiting pods that  belong to the same PodGroup. Otherwise, put the pod into WaitingPods and set the timeout (eg: 10s).
```